internal/controller/ipam/clusteripamclaim_controller.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2024
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipam
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ipamv1 "sigs.k8s.io/cluster-api/exp/ipam/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	kcm "github.com/K0rdent/kcm/api/v1alpha1"
+	"github.com/K0rdent/kcm/internal/utils"
+	"github.com/K0rdent/kcm/internal/utils/ratelimit"
+)
+
+type ClusterIPAMClaimReconciler struct {
+	client.Client
+	Scheme             *runtime.Scheme
+	defaultRequeueTime time.Duration
+}
+
+func (r *ClusterIPAMClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := ctrl.LoggerFrom(ctx)
+
+	ci := &kcm.ClusterIPAMClaim{}
+	if err := r.Client.Get(ctx, req.NamespacedName, ci); err != nil {
+		if apierrors.IsNotFound(err) {
+			l.Info("ClusterIPAMClaim not found, ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+
+		l.Error(err, "Failed to get ClusterIPAMClaim")
+		return ctrl.Result{}, err
+	}
+
+	ci.TypeMeta = metav1.TypeMeta{
+		APIVersion: kcm.GroupVersion.String(),
+		Kind:       "ClusterIPAMClaim",
+	}
```

Comment:
Why can't it be set in the tests rather than in the controller? Duplicating [this comment](https://github.com/k0rdent/kcm/pull/1260#discussion_r2016532270)